### ISSUE TITLE
Understanding how pipes / streams react to errors - weird issues

### ISSUE THREAD

**Issue by gannons** [nodejs/help#1979]
* **Node.js Version**: v10.15.3
* **OS**: Mac
* **Scope (install, code, runtime, meta, other?)**: runtime
* **Module (and version) (if relevant)**: stream, archiver

I'm using the following code to download and decrypt files.

      // Kick off an async promise to download the file
      promises.push(new Promise((resolve, reject) => {
        // Create a request for the file
        const uri = `http://127.0.0.1:8080/${file}`
        const req = request(encodeURI(uri)) 

        // Stream download the file
        const stream = req
          .on('error', (error) => {
            console.error(`Failure downloading ${file} with error ${error}`);
            // Always resolve an error. It should not stop the stream. But will be reported on the UI.
            resolve();
          })
          // Pipe it to decryption
          .pipe(decipher)
          .on('error', (error) => {
            console.error(`Error decrypting ${file} due to error ${error}`);
          })
          .on('finish', () => {
            console.timeEnd(file);
            resolve();
          });
          // .pipe(progressStream({drain: false}, (progress) => {}))
          // .pipe(passThrough.PassThrough())
          // .on('error', (error) => {
          //   console.error(`Error progressing ${file} due to error ${error}`);
          // });

        archive.append(stream, { name: file });
      }));

but I'm experiencing a few _weird_ scenarios I don't understand. 

First if there are no errors the above all works as expected. i.e. No request download or decryption errors.

But if there is an error

 1) The req on 'error' does not fire - instead it falls into the decipher on 'error'

 2) The decipher 'error' is passed onto archiver and the read stream to be _unpiped_.  i.e. The decipher 'error' causes both an 'error' and 'unpipe' event from the archiver. The 'unpipe' is a known result, as documented [here](https://blog.hustle.com/streams-and-pipes-and-errors-oh-my-62bc07b48ff1), if the writestream emits an error. But why would a 'decipher' error be passed onto the archiver?

 3) If I uncomment the code after 'finish' and pipe to **progressStream** or **passThrough** there is no 'unpipe' event on archiver. Instead the app hangs and the promise is never resolved. Seems like the 'unpipe' still happens but no event I can find is emitted.

While I somewhat understand 2) the other points confuse me. Why would piping to additional streams after the **decipher** hide the 'unpipe' event. Why does only the **decipher** report errors.

Any documentation or links explaining this would be appreciated.

**Comment by Hakerh400** [comment]
Can you post the full script? What is the `request` identifier pointing to? If it's `http.request` then that's not how you're supposed to use it. If it's the [request](https://www.npmjs.com/package/request) module from NPM, then please mention it somewhere. What is the `decipher`? You script is lacking some details. The `archiver` seems to be irrelevant here. Can you try to simplify your code and remove external dependencies (like `request` and `archiver`)?

**Comment by sregger** [comment]
@Hakerh400 I've created a repo to reproduce the issue

https://github.com/sregger/pipe-issue

> What is the request identifier pointing to?

It's the request module from npm. In the project above I'm pointing it at a locally run http server. But it could also be any file server. 

> What is the decipher?

The file's being downloaded are encrypted. Decipher is used to decrypt them. It does this using the built in **crypto** module.

> The archiver seems to be irrelevant here.

Archiver seems to be at the heart of the problem. When I'm seeing a decipher error I would expect it to be handled by an on('error'...) check. But instead the problem seems to migrate down to archiver. Here archiver reports an error and this, I believe, prompts the 'unpipe' event.

**Comment by Hakerh400** [comment]
Thank you for the clean repro steps now. Here's what's happening:

> 1. The req on 'error' does not fire - instead it falls into the decipher on 'error'
...
Once the testData is generated open testData/file_1.txt and modify the file. This'll cause the decryption to fail.

You basically answered your question. Request doesn't fire error event because there are no errors with the request itself. The file is on the disk and it is downloaded properly. If it can't be decrypted, request itself has nothing to do with it.

> The decipher 'error' is passed onto archiver and the read stream to be unpiped. i.e. The decipher 'error' causes both an 'error' and 'unpipe' event from the archiver. The 'unpipe' is a known result, as documented [here](https://blog.hustle.com/streams-and-pipes-and-errors-oh-my-62bc07b48ff1), if the writestream emits an error. But why would a 'decipher' error be passed onto the archiver?

What do you expect to happen instead? If an error occurs in the stream from which the archiver is reading, wouldn't the only reasonable thing to do be to report an error from withing the archiver? Please note that `archiver` doesn't internally do simple piping, instead it attaches listeners to the stream itself, including the `error` listener and they detect the error. That is not something instrinsic to Node.js streams, but simply an implementation decision of `archiver`'s developers.

> If I uncomment the code after 'finish' and pipe to progressStream or passThrough there is no 'unpipe' event on archiver. Instead the app hangs and the promise is never resolved. Seems like the 'unpipe' still happens but no event I can find is emitted.

Unpipe doesn't happen in that case, nor is it emitted. That's because the pass-through stream doesn't emit the error and that is the stream that you're passing to the archiver. The promise never resolves because `finish` event is not emitted (instead `error` event is emitted on the decipher stream). You should call `reject` (or `resolve`) in the error listener for the decipher stream.

---

To summarize, your questions are not related to Node.js streams, but rather to the implementation details of some external modules. I think it would be better idea to ask on their repositories why they decided to implement these functionalities that way, or you can suggest them better approach. In my opinion, how `archiver` currently works is the most reasonable implementation and I'm failing to see why do you think it should behave differently.

Hope this helps.

**Comment by sregger** [comment]
> You basically answered your question. Request doesn't fire error event because there are no errors with the request itself. The file is on the disk and it is downloaded properly. If it can't be decrypted, request itself has nothing to do with it.

My apologies that was misleading for point 1. In the README.md I said

> Once the testData is generated open testData/file_1.txt and modify the file. This'll cause the decryption to fail.

This is true and valid for points 2 and 3 in the original post. To reproduce point 1 you'll need to make the request "fail". To do so you can rename **testData/file_1.txt** to something else. e.g. **testData/file_X.txt **. Now the file does not exist, I thought, it'd generate an error. But it did not reported a request error. Instead I saw the decipher on 'error' fire. I did not understand why.

However I was being confused by the decipher / pipe / archive interaction. As a result when I noticed this problem I lumped them together as the same thing. But StackOverflow shows me this is because requestjs does not report an on 'error' if the file does not exist. e.g.

https://stackoverflow.com/questions/7222982/node-request-how-to-determine-if-an-error-occurred-during-the-request

**Consider point 1 explained.**

> What do you expect to happen instead? If an error occurs in the stream from which the archiver is reading, wouldn't the only reasonable thing to do be to report an error from withing the archiver? Please note that archiver doesn't internally do simple piping, instead it attaches listeners to the stream itself, including the error listener and they detect the error. That is not something instrinsic to Node.js streams, but simply an implementation decision of archiver's developers.

This may be the explanation I was looking for. In short I expected the on 'error' to handle the failure. i.e. Meaning the request to download the file would terminate. This does appear to happen. But the following may be what I misunderstood

> error occurs in the stream from which the archiver is reading

In my mind archiver was not **reading** from the stream. Instead it was being **written** to. But even if it **reads** from the stream I assumed the decipher's on 'error' would have _handled_ the error. i.e. Meaning archiver would not have the chance to read the error from the stream as it was already _dealt with_.

Basically I assumed the on 'error' did not propagate as appears to be the case with pipes. But, since you say, archiver does not use pipes that would explain why the behaviour is not the same.

> That is not something instrinsic to Node.js streams, but simply an implementation decision of archiver's developers.

Is there any documentation on error propagation with nodejs streams? i.e. I assume the stream has to pass the info onto archiver for it to behave in this manner.

**That handles point 2.**

> The promise never resolves because finish event is not emitted (instead error event is emitted on the decipher stream).

But the decipher's on 'finish' event runs. You can see this as

    console.timeEnd(file.name);

produces messages like

> file_1.txt: 48.496ms

i.e. Calculating how long it takes to download the file.

> Seems like the 'unpipe' still happens but no event I can find is emitted.

Let me correct myself. I've noticed there is an 'unpipe' event from the progressStream and passThrough. But re-piping it does not solve the issue.

This means that once the error occurs I can't correct the issue if I pipe anything after the decipher.

I also don't understand the following

        const req = request(encodeURI(decryptInfo.url))
        const ps = progressStream({drain: false}, (progress) => {})

        // Stream download the file
        const stream = req
          .on('error', (error) => {
            console.error(`Failure downloading ${file.name} with error ${error}`);
            // Always resolve an error. It should not stop the stream. But will be reported on the UI.
            resolve();
          })
          .on('unpipe', (error) => {
            console.error(`REQUEST unpipe event`);
          })
          // Pipe it to decryption
          .pipe(decipher)
          .on('error', (error) => {
            console.error(`Error decrypting ${file.name} due to error ${error}`);
          })
          .on('finish', () => {
            console.log('Finish called for ' + file.name);
            console.timeEnd(file.name);
            resolve();
          })
          .pipe(ps)
          .on('error', () => console.error('******Progress Error************'))
          .on('unpipe', (source) => {
            console.error('!!!!!!!!Progress Unpipe!!!!!!!!! ' + file.name)
            source.pipe(ps)
          });

with this code the output I see is

    [pipe-issue]$ node streamer.js 
    ********************************************************************************
    Downloading file_1.txt
    Downloading file_2.txt
    Error decrypting file_1.txt due to error Error: Unsupported state or unable to authenticate data
    Finish called for file_1.txt
    file_1.txt: 37.707ms
    Finish called for file_2.txt
    file_2.txt: 31.967ms
    !!!!!!!!Progress Unpipe!!!!!!!!! file_2.txt
    ********************************************************************************
    Downloading file_3.txt
    Downloading file_4.txt
    Finish called for file_3.txt
    file_3.txt: 5.822ms
    !!!!!!!!Progress Unpipe!!!!!!!!! file_3.txt
    [pipe-issue]$ 

Namely I don't get why an error on the file_1.txt stream produces an 'unpipe' event on the file_2.txt stream (which runs in parallel with file_1.txt) and then another 'unpipe' on file_3.txt's stream (which is not run in parallel with file_1.txt). 

I would have anticipated an 'unpipe' for file_1.txt with no impact to file_2.txt, file_3.txt, and file_4.txt. Instead 2 and 3 get 'unpipe' events and 4 does not 'finish'. i.e. Causing a "hang".

> Hope this helps.

It has been a big help.

**Comment by Hakerh400** [comment]
Answering remaining questions...

> Namely I don't get why an error on the file_1.txt stream produces an 'unpipe' event on the file_2.txt

Because you're using the same archiver. Once an error occurs, it ignores the remaining streams.

> which runs in parallel with file_1.txt

Request and decipher run in parallel, but archiver processes them sequentially. It doesn't start processing the second file until it finishes with the first file, so if an error occurs in the first file, the second file is "unpiped" instantly.

> But the decipher's on 'finish' event runs. You can see this as
...
Instead 2 and 3 get 'unpipe' events and 4 does not 'finish'. i.e. Causing a "hang"

You very nicely answered your question. The real question here, though, is *"why is `finish` emitted for the second and third file"* - I'm not sure, but probably has to do something with how `archiver` is implemented internally (again - it's not related to how Node.js streams work).

BTW, I just noticed there are two accounts here with the same avatar, am I talking with the same person?

**Comment by sregger** [comment]
> Because you're using the same archiver. Once an error occurs, it ignores the remaining streams.

In the above ProgressSteam scenario archiver reports no errors. Instead file_1.txt's decipher fires the only error. That appears to stop that pipe.

file_2.txt & file_3.txt both have separate pipes. All 3 eventually make there way to archiver. But before this point there should be no interaction.

Since archiver is quiet I thought an error might be reported in the ProgressStream. But it too seems error free. The only events I see are 'unpipe' on the 2 separate ProgressStreams.

I vaguely suspected it's somehow messing up Request or Decipher. But as they're separate instances that seems surprising.

> if an error occurs in the first file, the second file is "unpiped" instantly.

To me that makes sense when archiver is reporting an error. But in the ProgressStream scenario it is reporting itself as error free.

>  The real question here, though, is "why is finish emitted for the second and third file" 

AFAIK finish is emitted in error scenarios.

> BTW, I just noticed there are two accounts here with the same avatar, am I talking with the same person?

You are. Multiple accounts & multiple machines with different accounts signing in on each. Really don't need 2 accounts. It's a hold over. Plus merging them looks to be a pain.

**Comment by sregger** [comment]
This behavior seems specific to pipe. If I remove the archiver append, replace requestjs with a fs.createReadStream, force fs to fail by renaming the file then I see the same output (as I did by making each change in sequence and testing it).

      promises.push(new Promise((resolve, reject) => {
        // Create a request for the file
        // const req = request(encodeURI(decryptInfo.url))
        const req = fs.createReadStream(`./testData/${file.name}`);
        const ps = progressStream({drain: false}, (progress) => {})

        // Stream download the file
        const stream = req
          .on('error', (error) => {
            console.error(`Failure downloading ${file.name} with error ${error}`);
            // Always resolve an error. It should not stop the stream. But will be reported on the UI.
            resolve();
          })
          // Pipe it to decryption
          .pipe(decipher)
          .on('error', (error) => {
            console.error(`Error decrypting ${file.name} due to error ${error}`);
          })
          .on('finish', () => {
            console.timeEnd(file.name);
            resolve();
          })
          .pipe(ps)
          .on('error', () => console.error(`******Progress Error ${file.name}************`))
          .on('unpipe', (source) => {
            console.error('!!!!!!!!Progress Unpipe!!!!!!!!! ' + file.name)
            //source.pipe(ps)
          });

        //archive.append(stream, { name: file.name });
      }));

Produces

	C:\Users\gannons\Development\pipe-issue>node streamer.js
	********************************************************************************
	Downloading file_1.txt
	Downloading file_2.txt
	Failure downloading file_1.txt with error Error: ENOENT: no such file or directory, open 'C:\Users\gannons\Development\pipe-issue\testData\file_1.txt'
	file_2.txt: 20.216ms
	!!!!!!!!Progress Unpipe!!!!!!!!! file_2.txt
	********************************************************************************
	Downloading file_3.txt
	Downloading file_4.txt
	file_3.txt: 6.867ms
	!!!!!!!!Progress Unpipe!!!!!!!!! file_3.txt

I'm missing something fundamental. These pipes have no connection / merge point. AFAIK they're independent and working in parallel. Why an error in one would cause an 'unpipe' on another is unexplained.

The only way I've found to produce a change is to switch the drain property of the ProgressStream to true. This still produces 'unpipe' events (more even). But at least the pipes are all finished and finalize is invoked. e.g.

Changing

    const ps = progressStream({drain: false}, (progress) => {})

to

    const ps = progressStream({drain: true}, (progress) => {})

produces

	C:\Users\gannons\Development\pipe-issue>node streamer.js
	********************************************************************************
	Downloading file_1.txt
	Downloading file_2.txt
	Failure downloading file_1.txt with error Error: ENOENT: no such file or directory, open 'C:\Users\gannons\Development\pipe-issue\testData\file_1.txt'
	file_2.txt: 13.415ms
	!!!!!!!!Progress Unpipe!!!!!!!!! file_2.txt
	********************************************************************************
	Downloading file_3.txt
	Downloading file_4.txt
	file_3.txt: 10.679ms
	!!!!!!!!Progress Unpipe!!!!!!!!! file_3.txt
	file_4.txt: 10.296ms
	!!!!!!!!Progress Unpipe!!!!!!!!! file_4.txt
	********************************************************************************
	Downloading file_5.txt
	file_5.txt: 39.201ms
	!!!!!!!!Progress Unpipe!!!!!!!!! file_5.txt
	********************************************************************************
	Calling finalize
	++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	archive on unpipe
	++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	--------------------------------------------------------------------------------
	22 B total compressed bytes
	--------------------------------------------------------------------------------
	archive close called

Here you can see an archiver is being created but, as we no longer pipe anything to it, it is essentially empty at 22 Bytes. We can also see that the promises were all resolved as we managed to reach the finalize.

**Comment by sregger** [comment]
> I'm missing something fundamental. These pipes have no connection / merge point. AFAIK they're independent and working in parallel. Why an error in one would cause an 'unpipe' on another is unexplained.

Because, answering my own question, an 'unpipe' event is normal. It'll occur when all pipes run with no error. Basically when the read stream is finished it'll unpipe the write stream.

Kinda obvious when you realize it.

aka 'unpipe' in this scenario is a red herring.

That just leaves me confused over why, with drain: false, the script never reaches done.

    const ps = progressStream({drain: false}, (progress) => {})

I've uploaded [simple-stream](https://github.com/sregger/pipe-issue/blob/master/simple-stream.js) to highlight that.

**Comment by PoojaDurgad** [comment]
@gannons  , looks like this issue can be closed now. let me know if that is not the case.
